craft_application/services/state.py

```diff
@@ -0,0 +1,365 @@
+            if key in data and not overwrite:
+                raise ValueError(f"key {key!r} already exists and overwrite is false.")
+            data[key] = value
+            return
+
+        data_at_key = data.setdefault(key, {})
+        if not isinstance(data_at_key, dict):
+            raise KeyError(f"can't traverse into node at {key!r}.")
+
+        self._set(*remaining, data=data_at_key, value=value, overwrite=overwrite)
+
+    @final
+    def _destroy_state_dir(self) -> None:
+        """Remove the state directory.
+
+        If CRAFT_DEBUG is set, then the directory is never destroyed, even if the
+        application exits with an error.
+        """
+        if not os.getenv(CRAFT_DEBUG_ENV):
+            shutil.rmtree(self.__state_dir)
+
+    @final
+    def _create_state_dir(self) -> None:
+        """Create the state directory.
+
+        :raises StateServiceError: If the directory can't be created.
+        """
+        try:
+            self.__state_dir.mkdir(parents=True, exist_ok=True)
+        except FileExistsError as err:
+            raise errors.StateServiceError(
+                f"Failed to create state dir at {str(self.__state_dir)!r} because a "
+                "file with that name already exists."
+            ) from err
+
+    @final
+    @staticmethod
+    def _get_state_dir() -> pathlib.Path:
+        """Get the state directory.
+
+        Prefers, in order:
+        1. The CRAFT_STATE_DIR environment variable.
+        2. XDG_RUNTIME_DIR environment variable and a subdirectory with the outer instance's PID.
+        3. A temporary directory in the user's home directory.
+
+        When running in managed mode, the state directory must be set by CRAFT_STATE_DIR.
+
+        :raises StateServiceError: If the directory can't be determined.
+
+        :returns: The path to the state directory.
+        """
+        if state_dir_env := os.getenv(CRAFT_STATE_DIR_ENV):
+            state_dir = pathlib.Path(state_dir_env)
+            craft_cli.emit.debug(f"Getting state directory from {CRAFT_STATE_DIR_ENV}.")
+        elif util.is_managed_mode():
+            raise errors.StateServiceError(
+                f"Couldn't get state directory from {CRAFT_STATE_DIR_ENV} in managed mode."
+            )
+        else:
+            craft_cli.emit.debug("Getting state directory from XDG_RUNTIME_DIR.")
+            try:
+                # Using the PID as a subdir prevents conflicts when running
+                # multiple executions of the same application in parallel.
+                state_dir = pathlib.Path(BaseDirectory.get_runtime_dir()) / str(
+                    os.getpid()
+                )
+            except KeyError:
+                craft_cli.emit.debug("Couldn't get XDG_RUNTIME_DIR.")
+                # Use the home dir because the state dir will get mounted
+                # in the instance and Multipass can't access /tmp.
+                state_dir = util.get_home_temporary_directory()
```

Comment:
Since we still support macos, maybe it makes more sense to do something like:

```suggestion
            craft_cli.emit.debug("Getting runtime state directory.")
            # Using the PID as a subdir prevents conflicts when running
            # multiple executions of the same application in parallel.
            state_dir = platformdirs.user_runtime_path(os.getpid(), ensure_exists=True)
            if sys.platform == "linux" and state_dir.parents[-1] == "tmp":
                # Use the home dir because the state dir will get mounted
                # in the instance and Multipass can't access /tmp.
                state_dir = util.get_home_temporary_directory()
```

Reply:
Updated.